File: 卡码/代码随想录/1/11-图论/Go/dfs/LeetCode_0130.go
// 130. 被围绕的区域
package main

func solve(board [][]byte) {
	n, m := len(board), len(board[0])
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	var dfs func(x, y int)
	dfs = func(x, y int) {
		board[x][y] = 'Y'
		for i := 0; i < 4; i++ {
			x1 := x + dir[0][i]
			y1 := y + dir[1][i]
			if x1 < 0 || x1 > n-1 || y1 < 0 || y1 > m-1 {
				continue
			}
			if board[x1][y1] == 'O' {
				dfs(x1, y1)
			}
		}
	}
	for i := 0; i < n; i++ {
		if board[i][0] == 'O' {
			dfs(i, 0)
		}
		if board[i][m-1] == 'O' {
			dfs(i, m-1)
		}
	}
	for j := 0; j < m; j++ {
		if board[0][j] == 'O' {
			dfs(0, j)
		}
		if board[n-1][j] == 'O' {
			dfs(n-1, j)
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if board[i][j] == 'O' {
				board[i][j] = 'X'
			}
			if board[i][j] == 'Y' {
				board[i][j] = 'O'
			}
		}
	}
}

File: 卡码/代码随想录/1/11-图论/Python/dfs/LeetCode_0130.py
#130. 被围绕的区域
from typing import *

class Solution:
	def __init__(self):
		self.dir:List[List[int]]=[[0,1],[1,0],[0,-1],[-1,0]]
	def dfs(self,board: List[List[str]],x:int,y:int):
		board[x][y]="Y"
		for i in range(4):
			nextX:int=x+self.dir[i][0]
			nextY:int=y+self.dir[i][1]
			if nextX<0 or nextX>len(board)-1 or nextY<0 or nextY>len(board[0])-1:
				continue
			if board[nextX][nextY]=="O":
				self.dfs(board,nextX,nextY)
	def solve(self, board: List[List[str]]) -> None:
		n,m=len(board),len(board[0])
		for i in range(n):
			if board[i][0]=="O":
				self.dfs(board,i,0)
			if board[i][m-1]=="O":
				self.dfs(board,i,m-1)
		for j in range(m):
			if board[0][j]=="O":
				self.dfs(board,0,j)
			if board[n-1][j]=="O":
				self.dfs(board,n-1,j)
		for i in range(n):
			for j in range(m):
				if board[i][j]=="O":
					board[i][j]="X"
				if board[i][j]=="Y":
					board[i][j]="O"